### PR TITLE
HSB-473 feat: encrypt sensitive data before storing in db

### PR DESCRIPTION
**Pull request by mirarifhasan**
Closes HSB-473

This PR introduces data encryption before storing sensitive data in the database. For example, previously we stored SSO's secret key as plain text in the infraConfig table. From now on, before saving to DB, the value will be encrypted.
Also, after fetching the data from the DB, we decrypt it to the original value to use.

### What's changed

In `Account` and `InfraConfig` tables, we added encryption and decryption logic to store/ fetch sensitive data.

### Notes to reviewers

Nil

### CODE REVIEW

File: packages/hoppscotch-backend/src/infra-config/helper.ts [hunk 1]
```diff
@@ -76,119 +80,150 @@ export async function loadInfraConfiguration() {
...
     },
     {
       name: InfraConfigEnum.MAILER_TLS_REJECT_UNAUTHORIZED,
       value: process.env.MAILER_TLS_REJECT_UNAUTHORIZED,
+      isEncrypted: false,
     },
     {
       name: InfraConfigEnum.GOOGLE_CLIENT_ID,
       value: process.env.GOOGLE_CLIENT_ID,
+      isEncrypted: false,
```
Comment (balub) [inline packages/hoppscotch-backend/src/infra-config/helper.ts:146]: Lets encrypt this too.

File: packages/hoppscotch-backend/src/infra-config/helper.ts [hunk 2]
```diff
@@ -76,119 +80,150 @@ export async function loadInfraConfiguration() {
...
       name: InfraConfigEnum.ANALYTICS_USER_ID,
       value: generateAnalyticsUserId(),
+      isEncrypted: false,
     },
     {
       name: InfraConfigEnum.IS_FIRST_TIME_INFRA_SETUP,
       value: (await prisma.infraConfig.count()) === 0 ? 'true' : 'false',
+      isEncrypted: false,
     },
   ];
```
Comment (balub) [inline packages/hoppscotch-backend/src/infra-config/helper.ts:228]: Lets encrypt all the client ids too.

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts [hunk 3]
```diff
@@ -62,10 +65,30 @@ export class InfraConfigService implements OnModuleInit {
...
+      if (encryptionRequiredEntries.length > 0) {
+        const dbOperations = encryptionRequiredEntries.map((dbConfig) => {
+          return this.prisma.infraConfig.update({
+            where: { name: dbConfig.name },
+            data: { value: encrypt(dbConfig.value), isEncrypted: true },
+          });
+        });
+
+        await Promise.allSettled(dbOperations);
+      }
```
Comment (balub) [inline packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:88]: Can we generate an object that is essentially all the rows whose value needs to be encrypted and then just have one db query perform the operation on them all at once? Similiar to the import operations we have for collections.

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts [hunk 4]
```diff
@@ -118,9 +152,14 @@ export class InfraConfigService implements OnModuleInit {
     if (E.isLeft(isValidate)) return E.left(isValidate.left);
 
     try {
+      const { isEncrypted } = await this.prisma.infraConfig.findUnique({
+        where: { name },
+        select: { isEncrypted: true },
+      });
```
Comment (balub) [inline packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:158]: Does this DB call need to be made? can't we just get this info from `getDefaultInfraConfigs`.

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts [hunk 5]
```diff
@@ -146,11 +185,23 @@ export class InfraConfigService implements OnModuleInit {
     if (E.isLeft(isValidate)) return E.left(isValidate.left);
 
     try {
+      const dbInfraConfig = await this.prisma.infraConfig.findMany({
+        select: { name: true, isEncrypted: true },
+      });
```
Comment (balub) [inline packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:190]: Same here too, can't we use `getDefaultInfraConfigs` instead.

**Review by balub — CHANGES_REQUESTED** [review]

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts (hunk 3 above)
Comment (mirarifhasan) [reply packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:88]: That is not possible. In `updateMany` we can not write multiple where conditions.

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts (hunk 4 above)
Comment (mirarifhasan) [reply packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:158]: `getDefaultInfraConfigs` also has a DB (`count`) call. So, ultimately it doesn't help much.

File: packages/hoppscotch-backend/src/infra-config/infra-config.service.ts (hunk 5 above)
Comment (mirarifhasan) [reply packages/hoppscotch-backend/src/infra-config/infra-config.service.ts:190]: `getDefaultInfraConfigs` has a DB call.

File: packages/hoppscotch-backend/src/infra-config/helper.ts (hunk 1 above)
Comment (mirarifhasan) [reply packages/hoppscotch-backend/src/infra-config/helper.ts:146]: Ok

File: packages/hoppscotch-backend/src/infra-config/helper.ts [hunk 6]
```diff
@@ -76,119 +80,150 @@ export async function loadInfraConfiguration() {
...
+      isEncrypted: true,
     },
     {
       name: InfraConfigEnum.MAILER_TLS_REJECT_UNAUTHORIZED,
       value: process.env.MAILER_TLS_REJECT_UNAUTHORIZED,
+      isEncrypted: false,
     },
     {
       name: InfraConfigEnum.GOOGLE_CLIENT_ID,
       value: process.env.GOOGLE_CLIENT_ID,
```
Comment (balub) [inline packages/hoppscotch-backend/src/infra-config/helper.ts:145]:
```suggestion
      value: decrypt(process.env.GOOGLE_CLIENT_ID),
```

File: packages/hoppscotch-backend/src/infra-config/helper.ts [hunk 7]
```diff
@@ -76,119 +80,150 @@ export async function loadInfraConfiguration() {
...
+      isEncrypted: false,
     },
     {
       name: InfraConfigEnum.GOOGLE_SCOPE,
       value: process.env.GOOGLE_SCOPE,
+      isEncrypted: false,
     },
     {
       name: InfraConfigEnum.GITHUB_CLIENT_ID,
       value: process.env.GITHUB_CLIENT_ID,
```
Comment (balub) [inline packages/hoppscotch-backend/src/infra-config/helper.ts:165]:
```suggestion
      value: decrypt(process.env.GITHUB_CLIENT_ID),
```

File: packages/hoppscotch-backend/src/infra-config/helper.ts [hunk 8]
```diff
@@ -76,119 +80,150 @@ export async function loadInfraConfiguration() {
...
+      isEncrypted: false,
     },
     {
       name: InfraConfigEnum.GITHUB_SCOPE,
       value: process.env.GITHUB_SCOPE,
+      isEncrypted: false,
     },
     {
       name: InfraConfigEnum.MICROSOFT_CLIENT_ID,
       value: process.env.MICROSOFT_CLIENT_ID,
```
Comment (balub) [inline packages/hoppscotch-backend/src/infra-config/helper.ts:185]:
```suggestion
      value: decrypt(process.env.MICROSOFT_CLIENT_ID),
```

**Review by balub — CHANGES_REQUESTED** [review]

**Review by balub — APPROVED** [review]
LGTM 🚀

**Pull request merged**